src/mod_extdisco.erl

```diff
@@ -0,0 +1,136 @@
+        all_services ->
+            RequestedServices = get_external_services(),
+            {result, create_iq_response(RequestedServices)};
+        {credentials, {Type, Host}} ->
+            Services = get_external_services(Type),
+            RequestedServices = lists:filter(fun({_, Opts}) ->
+                gen_mod:get_opt(host, Opts, undefined) == binary_to_list(Host)
+                end, Services),
+            {result, create_iq_response_credentials(RequestedServices)};
+        {selected_services, Type} ->
+            RequestedServices = get_external_services(Type),
+            {result, create_iq_response_services(RequestedServices, Type)};
+        _ ->
+            {error, [mongoose_xmpp_errors:bad_request()]}
+    end,
+    {Acc, IQ#iq{type = ResponseType, sub_el = Response}}.
+
+request_type(#xmlel{name = <<"services">>} = Element) ->
+    case exml_query:attr(Element, <<"type">>) of
+        undefined -> all_services;
+        ServiceType ->
+            case catch binary_to_existing_atom(ServiceType, utf8) of
+                {'EXIT', _} -> {error, bad_request};
+                Type -> {selected_services, Type}
+            end
+    end;
+request_type(#xmlel{name = <<"credentials">>, children = [Children]}) ->
+    Host = exml_query:attr(Children, <<"host">>),
+    case exml_query:attr(Children, <<"type">>) of
+        undefined -> {error, bad_request};
+        ServiceType ->
+            case catch binary_to_existing_atom(ServiceType, utf8) of
+                {'EXIT', _} -> {error, bad_request};
+                Type -> {credentials, {Type, Host}}
+            end
+    end;
+request_type(_) ->
+    {error, bad_request}.
+
+create_iq_response(Services) ->
+    #xmlel{name = <<"services">>,
+        attrs = [{<<"xmlns">>, ?NS_EXTDISCO}],
+        children = prepare_services_element(Services, [])}.
+
+create_iq_response_services(Services, Type) ->
+    #xmlel{name = <<"services">>,
+        attrs = [{<<"xmlns">>, ?NS_EXTDISCO}, {<<"type">>, atom_to_binary(Type, utf8)}],
+        children = prepare_services_element(Services, [])}.
+
+create_iq_response_credentials(Services) ->
+    #xmlel{name = <<"credentials">>,
+        attrs = [{<<"xmlns">>, ?NS_EXTDISCO}],
+        children = prepare_services_element(Services, [])}.
+
+get_external_services() ->
+    case gen_mod:get_module_opts(?MYNAME, ?MODULE) of
+        [] -> [];
+        ServicesWithOpts -> ServicesWithOpts
+    end.
+
+get_external_services(Type) ->
+    [{T, Opts} || {T, Opts} <- get_external_services(), T == Type].
+
+prepare_services_element([], Result) ->
+    Result;
+prepare_services_element([{Type, Opts} | Services], Result) ->
+    RequiredElements = required_elements(Type, Opts),
+    OptionalElements = optional_elements(Opts),
+    NewResult = [#xmlel{name = <<"service">>,
+                        attrs = RequiredElements ++ OptionalElements}],
+    prepare_services_element(Services, Result ++ NewResult).
```

Comment:
Why don't you just fold? (foldl, since it should be faster)

Reply:
Could be a good idea. I think code would look like
```erlang
prepare_services_element(Services) ->
    lists:reverse(
      lists:foldl(
        fun({Type, Opts}, Acc) ->
                RequiredElements = required_elements(Type, Opts),
                OptionalElements = optional_elements(Opts),
                NewResult = [#xmlel{name = <<"service">>,
                                    attrs = RequiredElements ++ OptionalElements}],
                [NewResult | Acc]
        end, [], Services)).
```

Reply:
And with `lists:reverse/1` I'm just inverting the pattern where you do `Result ++ NewResult` in a more idiomatic way. I think that in that case, `Result` is a list with several elements, and `NewResult` is a single element, so the common pattern is to append a list of length `n` behind a single element, to keep the algorithm as `O(1)`, and then use the reverse bif that is very optimised for this. But I think that the order of the list doesn't make a difference when it comes to the XEP, but whatever 😄